dot/parachain/overseer/overseer.go

```diff
@@ -0,0 +1,115 @@
+// Copyright 2023 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package overseer
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/ChainSafe/gossamer/internal/log"
+)
+
+var (
+	logger = log.NewFromGlobal(log.AddContext("pkg", "parachain-overseer"))
+)
+
+type Overseer struct {
+	ctx                  context.Context
+	cancel               context.CancelFunc
+	errChan              chan error // channel for overseer to send errors to service that started it
+	SubsystemsToOverseer chan any
+	subsystems           map[Subsystem]chan any // map[Subsystem]OverseerToSubSystem channel
+	wg                   sync.WaitGroup
+}
+
+func NewOverseer() *Overseer {
+	ctx := context.Background()
+	ctx, cancel := context.WithCancel(ctx)
+	return &Overseer{
+		ctx:                  ctx,
+		cancel:               cancel,
+		errChan:              make(chan error),
+		SubsystemsToOverseer: make(chan any),
+		subsystems:           make(map[Subsystem]chan any),
+	}
+}
+
+// RegisterSubsystem registers a subsystem with the overseer,
+// Add OverseerToSubSystem channel to subsystem, which will be passed to subsystem's Run method.
+func (o *Overseer) RegisterSubsystem(subsystem Subsystem) {
+	o.subsystems[subsystem] = make(chan any)
+}
+
+func (o *Overseer) Start() error {
+	// start subsystems
+	for subsystem, overseerToSubSystem := range o.subsystems {
+		o.wg.Add(1)
+		go func(sub Subsystem, overseerToSubSystem chan any) {
+			err := sub.Run(o.ctx, overseerToSubSystem, o.SubsystemsToOverseer)
+			if err != nil {
+				logger.Errorf("running subsystem %v failed: %v", sub, err)
+			}
+			logger.Infof("subsystem %v stopped", sub)
+			o.wg.Done()
+		}(subsystem, overseerToSubSystem)
+	}
+
+	go o.processMessages()
+
+	// TODO: add logic to start listening for Block Imported events and Finalisation events
+	return nil
+}
+
+func (o *Overseer) processMessages() {
+	for {
+		select {
+		case msg := <-o.SubsystemsToOverseer:
+			switch msg.(type) {
+			default:
+				logger.Error("unknown message type")
+			}
```

Comment:
This is intentional. Actual message types haven't been added yet. Actual message types could be added along with subsystems. We can just copy message types from polkadot-sdk, but if we do it when we add the subsystem, the types would be more accurate since we would know what we need to send.
If we translate message types from polkadot-sdk, we might end up write a lot of varying data type code, unnecessarily.